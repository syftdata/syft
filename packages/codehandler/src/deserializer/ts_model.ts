import { Project, ScriptTarget } from 'ts-morph';
import { logDetail, logError, logFatal } from '@syftdata/common/lib/utils';
import {
  type InputSource,
  type AST,
  type Sink
} from '@syftdata/common/lib/types';
import { getEventSchemas } from './visitor';
import * as fs from 'fs';
import * as path from 'path';
import {
  getConfigExpressionForProject,
  getConfigFromExpression,
  getInputs,
  getSinks
} from './config';

// TODO: memoize this method.
export function createTSProject(filePaths: string[]): Project {
  const project = new Project({
    compilerOptions: { target: ScriptTarget.ES5, strictNullChecks: true }
  });

  // if directory, add glob style
  const files = filePaths.map((filePath) => {
    if (fs.existsSync(filePath) && fs.lstatSync(filePath).isDirectory()) {
      return path.join(filePath, '**', '!(*.d).ts');
    }
    return filePath;
  });
  if (files.length > 0) {
    project.addSourceFilesAtPaths(files);
  }
  return project;
}

export function readTSProject(
  sourceDir: string
): { project: Project; packageJson: any } | undefined {
  const tsConfigFilePath = `${sourceDir}/tsconfig.json`;
  if (!fs.existsSync(tsConfigFilePath)) {
    logError(
      `tsconfig.json not found in ${sourceDir}. Make sure you are in the root of a typescript project.`
    );
    return;
  }
  // read package.json file.
  const packageJsonPath = `${sourceDir}/package.json`;
  if (!fs.existsSync(packageJsonPath)) {
    logError(
      `package.json not found in ${sourceDir}. Make sure you are in the root of a typescript project.`
    );
    return;
  }
  // read all files in the project from source dir. and initialize ts-morph project
  const project = new Project({
    tsConfigFilePath
  });
  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
  return {
    project,
    packageJson
  };
}

export function generateASTForProject(project: Project): AST | undefined {
  const sourceFile = project.getSourceFile('events.ts');
  if (sourceFile == null) {
    logFatal(
      ':warning: events.ts file is not found. Please initialize your project using `syft init`'
    );
    return;
  }
  const relPath = path.relative('.', sourceFile.getFilePath());
  logDetail(`:heavy_check_mark: Parsing ${relPath}..`);
  const eventSchemas = getEventSchemas(project, sourceFile);

  const syftConfig = getConfigExpressionForProject(project);
  if (syftConfig === undefined) {
    logFatal(':warning: Config file is not found. Please provide one.');
    return;
  }
  const config = getConfigFromExpression(syftConfig);
  const sinks: Sink[] = getSinks(project);
  const inputs: InputSource[] = getInputs(project);
  return {
    eventSchemas,
    syftConfig,
    config,
    sinks,
    inputs
  };
}

export function deserialize(fileNames: string[]): AST | undefined {
  return generateASTForProject(createTSProject(fileNames));
}

export function deserializeFromCode(
  code: string,
  config: string
): AST | undefined {
  const project = new Project({
    compilerOptions: { target: ScriptTarget.ES5 }
  });
  project.createSourceFile(path.join('events.ts'), (writer) => {
    writer.write(code);
  });
  project.createSourceFile(path.join('config.ts'), (writer) => {
    writer.write(config);
  });
  return generateASTForProject(project);
}
