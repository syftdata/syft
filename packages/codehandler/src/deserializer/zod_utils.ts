import { type TypeField } from '@syftdata/common/lib/types';

const TypeToZodMapping = {
  number: 'number()',
  string: 'string()',
  boolean: 'boolean()',
  Date: 'date()',
  Object: 'object()'
};

const SyftTypeToZodMapping = {
  Email: 'string().email()',
  Url: 'string().url()',
  CountryCode: 'string().length(2)',
  UUID: 'string().uuid()',
  CUID: 'string().cuid()',
  CUID2: 'string().cuid2()'
};

export const ANY_TYPE = 'any()';

export function getZodType(
  type: string,
  enumValues: string[],
  syfttype?: string
): string {
  if (enumValues.length > 1) {
    // For Enum, return in the format of enum(["Salmon", "Tuna", "Trout"])
    return `enum([${enumValues.join(', ')}])`;
  }

  if (syfttype !== undefined) {
    return SyftTypeToZodMapping[syfttype] ?? ANY_TYPE;
  }
  return TypeToZodMapping[type] ?? ANY_TYPE;
}

export const ZOD_ALLOWED_TAGS: Record<string, string[]> = {
  number: [
    'min',
    'max',
    'positive',
    'negative',
    'nonnegative',
    'nonpositive',
    'multipleOf',
    'finite'
  ],
  string: ['min', 'max', 'length', 'startsWith', 'endsWith']
};

export function getZodTypeForSchema(fields: TypeField[]): string {
  return `z.object({
    ${fields
      .map((field) => `\t${field.name}: ${field.type.zodType}`)
      .join(',\n')}
  })`;
}
