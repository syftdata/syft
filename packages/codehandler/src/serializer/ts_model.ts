import { type AST } from '@syftdata/common/lib/types';
import * as handlebars from 'handlebars/dist/cjs/handlebars';
import { EVENT_MODELS_TEMPLATE } from '../templates/model_template';
import { SyftEventType } from '@syftdata/common/lib/client_types';
import { isEmptyString } from '@syftdata/common/lib/utils';

// TODO: Memoize
function getEventModelsTemplate(): handlebars.HandlebarsTemplateDelegate<any> {
  return handlebars.compile(EVENT_MODELS_TEMPLATE, { noEscape: true });
}

export function serialize(ast: AST): string {
  const schemas = ast.eventSchemas.map((schema) => ({
    ...schema,
    fields: schema.fields.map((field) => ({
      ...field,
      has_documentation: !isEmptyString(field.documentation)
    })),
    eventType: SyftEventType[schema.eventType]
  }));
  return getEventModelsTemplate()({
    schemas
  });
}
