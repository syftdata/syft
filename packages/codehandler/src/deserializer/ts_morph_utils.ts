import {
  type JSDoc,
  type JSDocTag,
  type ts,
  type Type,
  type Symbol as TSSymbol,
  SyntaxKind
} from 'ts-morph';
import { logInfo } from '@syftdata/common/lib/utils';
import {
  type TypeField,
  type Field,
  type TypeSchema
} from '@syftdata/common/lib/types';
import { ANY_TYPE, getZodType, getZodTypeForSchema } from './zod_utils';

// const SyftypeModuleIndex = 'client/src/index".type.'
const SyftypeIndex = 'type.';

export function getTags(docs: JSDoc[]): Array<JSDocTag<ts.JSDocTag>> {
  return docs.reduce<Array<JSDocTag<ts.JSDocTag>>>((prev, curr) => {
    return [...prev, ...curr.getTags()];
  }, []);
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function printType(type: Type): void {
  console.log('>>>>> printType ', type.getText());
  const proto = Object.getPrototypeOf(type);
  for (const member of Object.getOwnPropertyNames(proto)) {
    if (member.startsWith('is')) {
      if (typeof type[member] === 'function') {
        const result = type[member]();
        if (result === true) {
          console.log('>>> member ', member);
        }
      }
    }
  }
}

/**
 * This method returns TypeSchema of a field.
 * @param property
 * @returns
 */
function getTypeField(property: TSSymbol, debugName: string): TypeField {
  const declaration = property.getValueDeclarationOrThrow();
  const type = getTypeSchema(declaration.getType(), debugName);
  const hasQuestion =
    declaration.getFirstChildByKind(SyntaxKind.QuestionToken) != null;
  return {
    name: property.getName(),
    type,
    isOptional: hasQuestion
  };
}

/**
 * This method returns TypeSchema for complex objects like
 * class / interface / type.
 * @param typeObj
 * @returns
 */
function getTypeSchemaForComplexObject(
  typeObj: Type<ts.Type>,
  debugName: string
): TypeSchema {
  let name = typeObj.getSymbol()?.getEscapedName() ?? 'Unknown';
  if (typeObj.isAnonymous()) {
    name = typeObj.getAliasSymbol()?.getEscapedName() ?? name;
  }

  const typeFields = typeObj
    .getApparentProperties()
    .map((property) =>
      getTypeField(property, `${debugName}.${property.getEscapedName()}`)
    )
    .filter((field) => field != null) as Field[];

  const zodType = getZodTypeForSchema(typeFields);
  return {
    name,
    typeFields,
    zodType
  };
}

const KNOWN_TYPES = new Set(['Date', 'string', 'number', 'Object']);

/**
 * This method returns type-schema for all types of fields.
 * including primitives, enums, class/interface and types.
 * @param typeObj
 * @returns
 */
export function getTypeSchema(
  typeObj: Type<ts.Type>,
  debugName: string
): TypeSchema {
  let name = typeObj.getText();
  let enumValues: string[] = [];
  const isOptional = typeObj.isNullable();
  let syfttype: string | undefined;
  let foundUnsuppotedCloudType = false;
  if (typeObj.isClassOrInterface() || name.includes(SyftypeIndex)) {
    if (name.includes(SyftypeIndex)) {
      // TODO: all types that start with "type." are treated as syft-types.
      // Downside: if syft-types are imported as something else, this doesn't work.
      // if something else is imported as type. it gets aggected.
      const typeStrLoc = name.indexOf(SyftypeIndex);
      syfttype = name.slice(typeStrLoc + SyftypeIndex.length);
      name = 'string';
    } else if (!KNOWN_TYPES.has(name)) {
      foundUnsuppotedCloudType = true;
      return getTypeSchemaForComplexObject(typeObj, debugName);
    }
  } else if (typeObj.isObject()) {
    foundUnsuppotedCloudType = true;
    return getTypeSchemaForComplexObject(typeObj, debugName);
  } else if (typeObj.isUnion() && !typeObj.isBoolean()) {
    const subtypes = typeObj.getUnionTypes();
    let incompatibleSubType = false;
    enumValues = subtypes.map((subtype) => {
      if (subtype.isEnumLiteral()) {
        const enumVal = subtype.getLiteralValue()?.toString();
        if (enumVal != null) {
          return subtype.isStringLiteral() ? `"${enumVal}"` : enumVal;
        }
      } else if (
        subtype.isStringLiteral() ||
        subtype.isNumberLiteral() ||
        subtype.isBooleanLiteral()
      ) {
        return subtype.getText();
      }
      incompatibleSubType = true;
      return '';
    });
    if (!incompatibleSubType) {
      name = enumValues.join(' | ');
    } else {
      logInfo(
        `:warning: Unknown type is seen: "${debugName}: ${name}". using "any" for now`
      );
      name = 'any';
    }
  } else if (typeObj.isAny()) {
    // throw error.
    logInfo(`:warning: Any type is seen: "${debugName}".`);
    name = 'any';
    foundUnsuppotedCloudType = true;
  }

  if (foundUnsuppotedCloudType) {
    logInfo(
      `:prohibited: ${debugName} field is defined as ${name} type, which is not supported on the cloud.`
    );
  }

  let zodType = `z.${getZodType(name, enumValues, syfttype)}`;
  if (zodType === ANY_TYPE) {
    logInfo(
      `Used an unknown field type ${name}. No validations will be performed`
    );
  }

  if (isOptional) {
    zodType = `${zodType}.optional()`;
  }

  return {
    name,
    syfttype,
    zodType
  };
}
