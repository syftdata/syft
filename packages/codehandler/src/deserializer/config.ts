import {
  type StaticConfig,
  SyftSinkType,
  SyftInputType
} from '@syftdata/common/lib/client_types';
import { type InputSource, type Sink } from '@syftdata/common/lib/types';
import type { ArrayLiteralExpression, Project } from 'ts-morph';
import { SyntaxKind, ObjectLiteralExpression } from 'ts-morph';
import { getConfigObject } from './ts_morph_utils';

const STATIC_CONFIG_TYPE = 'StaticConfig';

export function getConfigExpressionForProject(
  project: Project
): ObjectLiteralExpression | undefined {
  let syftConfig: ObjectLiteralExpression | undefined;
  for (const sourceFile of project.getSourceFiles()) {
    sourceFile.getVariableDeclarations().forEach((variable) => {
      let varialbeType = variable.getType().getText();
      const symbol = variable.getType().getSymbol();
      if (symbol !== undefined) {
        varialbeType = symbol.getEscapedName();
      }
      if (varialbeType !== STATIC_CONFIG_TYPE) return;
      syftConfig = variable.getInitializerIfKind(
        SyntaxKind.ObjectLiteralExpression
      );
    });
    if (syftConfig !== undefined) break;
  }
  return syftConfig;
}

export function getConfigFromExpression(
  expression: ObjectLiteralExpression
): StaticConfig {
  const val: StaticConfig = {
    projectName: 'test',
    version: '0.0.0'
  };
  const config = getConfigObject(expression);
  if (config === undefined) return val;
  return {
    ...val,
    ...config
  };
}

function getArrayLiteralExpression(
  project: Project,
  name: string
): ArrayLiteralExpression | undefined {
  let sinks: ArrayLiteralExpression | undefined;
  for (const sourceFile of project.getSourceFiles()) {
    const arrays = sourceFile
      .getVariableDeclarations()
      .map((variable) => {
        if (variable.getName() === name) {
          return variable.getInitializerIfKind(
            SyntaxKind.ArrayLiteralExpression
          );
        }
        return undefined;
      })
      .filter((x) => x !== undefined) as ArrayLiteralExpression[];
    if (arrays.length > 0) {
      sinks = arrays[0];
    }
    if (sinks !== undefined) break;
  }
  return sinks;
}

export function getSinks(project: Project): Sink[] {
  const sinks = getArrayLiteralExpression(project, 'sinks');
  if (sinks === undefined) return [];
  const sinkObjects = sinks
    .getElements()
    .map((element) => {
      if (element instanceof ObjectLiteralExpression) {
        return getConfigObject(element) as unknown as Sink;
      }
      return undefined;
    })
    .filter((x) => x !== undefined) as Sink[];
  return sinkObjects;
}

export function getInputs(project: Project): InputSource[] {
  const inputs = getArrayLiteralExpression(project, 'inputs');
  if (inputs === undefined) return [];
  const inputObjects = inputs
    .getElements()
    .map((element) => {
      if (element instanceof ObjectLiteralExpression) {
        return getConfigObject(element) as unknown as InputSource;
      }
      return undefined;
    })
    .filter((x) => x !== undefined) as InputSource[];
  return inputObjects;
}
