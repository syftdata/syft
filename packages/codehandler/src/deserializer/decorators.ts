import { JSDocTypeTag, type Decorator, type JSDoc, SyntaxKind } from 'ts-morph';
import { type TypeField } from '@syftdata/common/lib/types';
import { SyftEventType } from '@syftdata/common/lib/client_types';
import { logError } from '@syftdata/common/lib/utils';
import { extractKVPairsFromObjectLiteral, getTags } from './ts_morph_utils';
import { ZOD_ALLOWED_TAGS } from './zod_utils';
import {
  type DBEventSource,
  type DBFieldRelation
} from '@syftdata/common/lib/db_types';

export interface EventProperties {
  eventType?: SyftEventType;
  dbSourceDetails?: DBEventSource;
}

export interface FieldProperties {
  zodType?: string;
  dbRelation?: DBFieldRelation;
  rename?: string;
}

export function extractFieldProperties(
  typeField: TypeField,
  decorators: Decorator[],
  docs: JSDoc[]
): FieldProperties {
  const tags = getTags(docs);
  let fieldRelationProperties: DBFieldRelation | undefined;
  let rename: string | undefined;
  decorators.forEach((decorator) => {
    const name = decorator.getName();
    if (name === 'relation') {
      const args = decorator.getArguments();
      if (args.length === 0) {
        logError(`@relation decorator requires an argument`);
        return;
      }
      const dbArg = args[0];
      if (dbArg.isKind(SyntaxKind.ObjectLiteralExpression)) {
        const kvPairs = extractKVPairsFromObjectLiteral(dbArg);
        if (
          kvPairs.has('table') &&
          kvPairs.has('references') &&
          kvPairs.has('fields')
        ) {
          fieldRelationProperties = {
            table: kvPairs.get('table') as string,
            references: kvPairs.get('references') as string[],
            fields: kvPairs.get('fields') as string[],
            isMany: typeField.type.isArray ?? false
          };
        }
      }
      if (fieldRelationProperties === undefined) {
        logError(
          `relation decorator requires table, references, fields properties`
        );
      }
    } else if (name === 'rename') {
      const args = decorator.getArguments();
      if (args.length === 0) {
        logError(`@rename decorator requires an argument`);
        return;
      }
      const dbArg = args[0];
      if (dbArg.isKind(SyntaxKind.StringLiteral)) {
        rename = dbArg.getLiteralValue();
      }
    }
  });

  let zodType = typeField.type.zodType;
  const allowedZodTags = ZOD_ALLOWED_TAGS[typeField.name] ?? [];
  tags.forEach((tag) => {
    const name = tag.getTagName();
    if (allowedZodTags.includes(name)) {
      const paramString = tag.getCommentText()?.trim();
      if (paramString !== undefined && paramString.length > 0) {
        if (name === 'startsWith' || name === 'endsWith') {
          zodType = `${zodType}.${name}("${paramString}")`;
        } else {
          const paramInt = parseInt(paramString);
          zodType = `${zodType}.${name}(${!isNaN(paramInt) ? paramInt : ''})`;
        }
      }
    }
  });
  return {
    zodType,
    dbRelation: fieldRelationProperties,
    rename
  };
}

export function extractEventProperties(
  decorators: Decorator[],
  docs: JSDoc[]
): EventProperties {
  const tags = getTags(docs);
  let eventType: SyftEventType | undefined;
  let dbProperties: DBEventSource | undefined;

  decorators.forEach((decorator) => {
    const name = decorator.getName();
    if (name === 'eventtype') {
      const args = decorator.getArguments();
      if (args.length === 0) {
        logError(`@eventtype decorator requires arguments`);
        return;
      }
      const arg = args[0];
      const eventTypeString = arg.getText();
      if (!eventTypeString.startsWith('SyftEventType.')) {
        logError(`@eventtype decorator requires SyftEventType enum`);
        return;
      }
      eventType = SyftEventType[eventTypeString.replace('SyftEventType.', '')];
      if (eventType === SyftEventType.DB) {
        const dbArg = args[1];
        if (dbArg.isKind(SyntaxKind.ObjectLiteralExpression)) {
          const kvPairs = extractKVPairsFromObjectLiteral(dbArg);
          if (kvPairs.has('table') && kvPairs.has('on')) {
            dbProperties = {
              table: kvPairs.get('table') as string,
              on: kvPairs.get('on') as string,
              fields: kvPairs.get('fields') as string[]
            };
          }
        }
        if (dbProperties === undefined) {
          logError(
            `eventtype decorator with SyftEventType.DB requires table and on properties`
          );
        }
      }
    }
  });

  tags.forEach((tag) => {
    if (tag instanceof JSDocTypeTag || tag.getTagName() === 'type') {
      const fullTypeString = (tag as JSDocTypeTag)
        .getTypeExpression()
        ?.getTypeNode()
        ?.getText();
      if (fullTypeString === undefined) return;
      const typeString = fullTypeString.replace('SyftEventType.', '');
      if (typeString in SyftEventType) {
        eventType = SyftEventType[typeString as keyof typeof SyftEventType];
      }
    }
  });
  return { eventType, dbSourceDetails: dbProperties };
}
