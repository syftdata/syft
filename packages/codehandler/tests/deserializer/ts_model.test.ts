/* eslint-disable @typescript-eslint/no-unused-vars */
import { deserialize } from '../../src/deserializer/ts_model';

describe('generateASTForProject', () => {
  it('with event annotations', async () => {
    const ast = deserialize(['./tests/test_schema/single_event']);

    expect(ast).toBeDefined();
    expect(ast?.eventSchemas.length).toBe(1);
    expect(JSON.stringify(ast?.eventSchemas, null, 2)).toMatchSnapshot();
  });

  it('with multiple schemas', async () => {
    const ast = deserialize(['./tests/test_schema/multiple_events']);
    expect(ast).toBeDefined();
    expect(ast?.eventSchemas.length).toBe(8);
    expect(JSON.stringify(ast?.eventSchemas, null, 2)).toMatchSnapshot();
  });
});
