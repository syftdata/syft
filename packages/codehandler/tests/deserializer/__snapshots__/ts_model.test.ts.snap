// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateASTForProject with event annotations 1`] = `
"[
  {
    "name": "UserIdentity",
    "eventType": 3,
    "exported": true,
    "documentation": "User Identity is tracked via this event.",
    "fields": [
      {
        "name": "userId",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "UUID",
          "zodType": "z.string().uuid()",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "email",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "Email",
          "zodType": "z.string().email()",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "country",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "CountryCode",
          "zodType": "z.string().length(2)",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "direction",
        "isOptional": false,
        "type": {
          "name": "number",
          "zodType": "z.enum([0, 1, 2, 3])",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "isSpammer",
        "isOptional": false,
        "type": {
          "name": "string",
          "zodType": "z.enum([\\"Yes\\", \\"No\\"])",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "isVerified",
        "isOptional": false,
        "type": {
          "name": "boolean",
          "zodType": "z.boolean()",
          "isArray": false
        },
        "documentation": ""
      }
    ],
    "zodType": "z.object({\\n    \\tuserId: z.string().uuid(),\\n\\temail: z.string().email(),\\n\\tcountry: z.string().length(2),\\n\\tdirection: z.enum([0, 1, 2, 3]),\\n\\tisSpammer: z.enum([\\"Yes\\", \\"No\\"]),\\n\\tisVerified: z.boolean()\\n  })"
  }
]"
`;

exports[`generateASTForProject with multiple schemas 1`] = `
"[
  {
    "name": "UserIdentity",
    "eventType": 3,
    "exported": true,
    "documentation": "User Identity is tracked via this event.",
    "fields": [
      {
        "name": "userId",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "UUID",
          "zodType": "z.string().uuid()",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "email",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "Email",
          "zodType": "z.string().email()",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "country",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "CountryCode",
          "zodType": "z.string().length(2)",
          "isArray": false
        },
        "documentation": ""
      }
    ],
    "zodType": "z.object({\\n    \\tuserId: z.string().uuid(),\\n\\temail: z.string().email(),\\n\\tcountry: z.string().length(2)\\n  })"
  },
  {
    "name": "BasePageEvent",
    "eventType": 1,
    "exported": false,
    "documentation": "@type {SyftEventType.PAGE}",
    "fields": [
      {
        "name": "page",
        "isOptional": false,
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "time",
        "isOptional": false,
        "type": {
          "name": "Date",
          "zodType": "z.date()",
          "isArray": false
        },
        "documentation": ""
      }
    ],
    "zodType": "z.object({\\n    \\tpage: z.string(),\\n\\ttime: z.date()\\n  })"
  },
  {
    "name": "IndexPageEvent",
    "eventType": 1,
    "exported": true,
    "documentation": "Gets logged when user opens the index page.",
    "fields": [
      {
        "name": "page",
        "isOptional": false,
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "documentation": "",
        "defaultValue": "'index'"
      },
      {
        "name": "loggedIn",
        "isOptional": false,
        "type": {
          "name": "boolean",
          "zodType": "z.boolean()",
          "isArray": false
        },
        "documentation": "",
        "defaultValue": "false"
      },
      {
        "name": "time",
        "isOptional": false,
        "type": {
          "name": "Date",
          "zodType": "z.date()",
          "isArray": false
        },
        "documentation": ""
      }
    ],
    "zodType": "z.object({\\n    \\tpage: z.string(),\\n\\tloggedIn: z.boolean(),\\n\\ttime: z.date()\\n  })"
  },
  {
    "name": "SourceType",
    "eventType": 0,
    "exported": false,
    "documentation": "",
    "fields": [
      {
        "name": "source_name",
        "isOptional": false,
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "source_id",
        "isOptional": true,
        "type": {
          "name": "number",
          "zodType": "z.number().optional()",
          "isArray": false
        },
        "documentation": ""
      }
    ],
    "zodType": "z.object({\\n    \\tsource_name: z.string(),\\n\\tsource_id: z.number().optional()\\n  })"
  },
  {
    "name": "SourceAction",
    "eventType": 0,
    "exported": false,
    "documentation": "SourceAction gets logged when a source is created/updated/deleted.",
    "fields": [
      {
        "name": "isOriginal",
        "isOptional": false,
        "type": {
          "name": "boolean",
          "zodType": "z.boolean()",
          "isArray": false
        },
        "documentation": "is the source original or a proxy."
      },
      {
        "name": "id",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "UUID",
          "zodType": "z.string().uuid()",
          "isArray": false
        },
        "documentation": "id of the source."
      },
      {
        "name": "type",
        "isOptional": false,
        "type": {
          "name": "SourceType",
          "typeFields": [
            {
              "name": "source_name",
              "isOptional": false,
              "type": {
                "name": "string",
                "zodType": "z.string()",
                "isArray": false
              },
              "documentation": ""
            },
            {
              "name": "source_id",
              "isOptional": true,
              "type": {
                "name": "number",
                "zodType": "z.number().optional()",
                "isArray": false
              },
              "documentation": ""
            }
          ],
          "zodType": "z.object({\\n    \\tsource_name: z.string(),\\n\\tsource_id: z.number().optional()\\n  })",
          "isArray": false
        },
        "documentation": "Source type."
      },
      {
        "name": "eventType",
        "isOptional": true,
        "type": {
          "name": "EventType",
          "typeFields": [
            {
              "name": "name",
              "isOptional": false,
              "type": {
                "name": "string",
                "zodType": "z.string()",
                "isArray": false
              },
              "documentation": ""
            },
            {
              "name": "id",
              "isOptional": true,
              "type": {
                "name": "number",
                "zodType": "z.number().optional()",
                "isArray": false
              },
              "documentation": ""
            },
            {
              "name": "test_field",
              "isOptional": true,
              "type": {
                "name": "string",
                "zodType": "z.string().optional()",
                "isArray": false
              },
              "documentation": ""
            }
          ],
          "zodType": "z.object({\\n    \\tname: z.string(),\\n\\tid: z.number().optional(),\\n\\ttest_field: z.string().optional()\\n  }).optional()",
          "isArray": false
        },
        "documentation": ""
      }
    ],
    "zodType": "z.object({\\n    \\tisOriginal: z.boolean(),\\n\\tid: z.string().uuid(),\\n\\ttype: z.object({\\n    \\tsource_name: z.string(),\\n\\tsource_id: z.number().optional()\\n  }),\\n\\teventType: z.object({\\n    \\tname: z.string(),\\n\\tid: z.number().optional(),\\n\\ttest_field: z.string().optional()\\n  }).optional()\\n  })"
  },
  {
    "name": "SourceUpdated",
    "eventType": 0,
    "exported": true,
    "documentation": "SourceAction gets logged when a source is created/updated/deleted.",
    "fields": [
      {
        "name": "isOriginal",
        "isOptional": false,
        "type": {
          "name": "boolean",
          "zodType": "z.boolean()",
          "isArray": false
        },
        "documentation": "is the source original or a proxy."
      },
      {
        "name": "id",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "UUID",
          "zodType": "z.string().uuid()",
          "isArray": false
        },
        "documentation": "id of the source."
      },
      {
        "name": "type",
        "isOptional": false,
        "type": {
          "name": "SourceType",
          "typeFields": [
            {
              "name": "source_name",
              "isOptional": false,
              "type": {
                "name": "string",
                "zodType": "z.string()",
                "isArray": false
              },
              "documentation": ""
            },
            {
              "name": "source_id",
              "isOptional": true,
              "type": {
                "name": "number",
                "zodType": "z.number().optional()",
                "isArray": false
              },
              "documentation": ""
            }
          ],
          "zodType": "z.object({\\n    \\tsource_name: z.string(),\\n\\tsource_id: z.number().optional()\\n  })",
          "isArray": false
        },
        "documentation": "Source type."
      },
      {
        "name": "eventType",
        "isOptional": true,
        "type": {
          "name": "EventType",
          "typeFields": [
            {
              "name": "name",
              "isOptional": false,
              "type": {
                "name": "string",
                "zodType": "z.string()",
                "isArray": false
              },
              "documentation": ""
            },
            {
              "name": "id",
              "isOptional": true,
              "type": {
                "name": "number",
                "zodType": "z.number().optional()",
                "isArray": false
              },
              "documentation": ""
            },
            {
              "name": "test_field",
              "isOptional": true,
              "type": {
                "name": "string",
                "zodType": "z.string().optional()",
                "isArray": false
              },
              "documentation": ""
            }
          ],
          "zodType": "z.object({\\n    \\tname: z.string(),\\n\\tid: z.number().optional(),\\n\\ttest_field: z.string().optional()\\n  }).optional()",
          "isArray": false
        },
        "documentation": ""
      }
    ],
    "zodType": "z.object({\\n    \\tisOriginal: z.boolean(),\\n\\tid: z.string().uuid(),\\n\\ttype: z.object({\\n    \\tsource_name: z.string(),\\n\\tsource_id: z.number().optional()\\n  }),\\n\\teventType: z.object({\\n    \\tname: z.string(),\\n\\tid: z.number().optional(),\\n\\ttest_field: z.string().optional()\\n  }).optional()\\n  })"
  },
  {
    "name": "NotifChannelAction",
    "eventType": 0,
    "exported": false,
    "documentation": "",
    "fields": [
      {
        "name": "name",
        "isOptional": false,
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "documentation": "Name of the channel, this is a unique global name.\\n@min 5"
      },
      {
        "name": "type",
        "isOptional": false,
        "type": {
          "name": "ChannelType",
          "typeFields": [
            {
              "name": "channel_name",
              "isOptional": false,
              "type": {
                "name": "string",
                "zodType": "z.string()",
                "isArray": false
              },
              "documentation": ""
            },
            {
              "name": "channel_type",
              "isOptional": false,
              "type": {
                "name": "string",
                "zodType": "z.string()",
                "isArray": false
              },
              "documentation": ""
            }
          ],
          "zodType": "z.object({\\n    \\tchannel_name: z.string(),\\n\\tchannel_type: z.string()\\n  })",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "webhook",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "Url",
          "zodType": "z.string().url()",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "count",
        "isOptional": false,
        "type": {
          "name": "number",
          "zodType": "z.number()",
          "isArray": false
        },
        "documentation": "Number of notifications fired so far.\\n@min 0"
      }
    ],
    "zodType": "z.object({\\n    \\tname: z.string(),\\n\\ttype: z.object({\\n    \\tchannel_name: z.string(),\\n\\tchannel_type: z.string()\\n  }),\\n\\twebhook: z.string().url(),\\n\\tcount: z.number()\\n  })"
  },
  {
    "name": "NotifChannelUpdated",
    "eventType": 0,
    "exported": true,
    "documentation": "",
    "fields": [
      {
        "name": "name",
        "isOptional": false,
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "documentation": "Name of the channel, this is a unique global name.\\n@min 5"
      },
      {
        "name": "type",
        "isOptional": false,
        "type": {
          "name": "ChannelType",
          "typeFields": [
            {
              "name": "channel_name",
              "isOptional": false,
              "type": {
                "name": "string",
                "zodType": "z.string()",
                "isArray": false
              },
              "documentation": ""
            },
            {
              "name": "channel_type",
              "isOptional": false,
              "type": {
                "name": "string",
                "zodType": "z.string()",
                "isArray": false
              },
              "documentation": ""
            }
          ],
          "zodType": "z.object({\\n    \\tchannel_name: z.string(),\\n\\tchannel_type: z.string()\\n  })",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "webhook",
        "isOptional": false,
        "type": {
          "name": "string",
          "syfttype": "Url",
          "zodType": "z.string().url()",
          "isArray": false
        },
        "documentation": ""
      },
      {
        "name": "count",
        "isOptional": false,
        "type": {
          "name": "number",
          "zodType": "z.number()",
          "isArray": false
        },
        "documentation": "Number of notifications fired so far.\\n@min 0"
      }
    ],
    "zodType": "z.object({\\n    \\tname: z.string(),\\n\\ttype: z.object({\\n    \\tchannel_name: z.string(),\\n\\tchannel_type: z.string()\\n  }),\\n\\twebhook: z.string().url(),\\n\\tcount: z.number()\\n  })"
  }
]"
`;
